### ISSUE TITLE
Dex doesn't retry when database not available during start

### ISSUE THREAD

**Issue by kaushikjeet** [dexidp/dex#3602]
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/dexidp/dex/blob/master/.github/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/dexidp/dex/issues) for an issue that matches the one I want to file, without success.
- [X] I am not looking for support or already pursued the available [support channels](https://github.com/dexidp/dex/issues/new/choose) without success.

### Version

2.32.0

### Storage Type

Postgres

### Installation Type

Binary, Custom container image

### Expected Behavior

Dex should retry database connections instead failing to start if database not available. In container world its difficult to align pods startup sequence in distributed environment. 

### Actual Behavior

Dex failed to start if unable to connect with Postgres during start time. It error out with message "failed to initialize storage:"

### Steps To Reproduce

1. configure postgres connector with Dex.
2. start Dex without bringing databases first.
3. Dex will fail to start with error "failed to initialize storage:"

### Additional Information

We have applications deployed over K3S, Dex is one of the pod having Postgres database connector. It seems to be mandatory have database available for Dex to start as it try to connect with Postgres and error out. It doesn't have logic to retry connection when failed.

### Logs

```shell
failed to initialize storage:
```

**Comment by nabokihms** [comment]
@kaushikjeet Yes, I believe there are many cases when Dex does not start because of external dependencies, such as other providers or databases. This is unwanted behavior, and it would be nice to have this issue fixed.

**Comment by kaushikjeet** [comment]
@nabokihms , shall we try fix and create PR for this behaviour?

**Comment by nabokihms** [comment]
Yeah, sure, go ahead. I believe this is a good improvement.

**Comment by MoeBensu** [comment]
@nabokihms I put something together. Did some manual testing and it looks so far promising. Still need to add proper tests, but I'm new to Go testing and Dex's setup, so that's coming next.

If you/anyone want to take a peek. Since I'm still familiarizing myself with Go, any feedback is appreciated!
